### PR TITLE
feat: added gdal logic

### PR DESCRIPTION
**Pull request by CL-SHLOMIKONCHA**
| Question                | Answer                                                                          |
| ---------------- | -------------------------------------------------------------------------- |
| Bug fix         | ✔/✖                                                                        |
| New feature     | ✔/✖                                                                        |
| Breaking change | ✔/✖                                                                        |
| Deprecations    | ✔/✖                                                                        |
| Documentation   | ✔/✖                                                                        |
| Tests added     | ✔/✖                                                                        |
| Chore            | ✔/✖                                                                       |

Related issues: #XXX , #XXX ...
Closes #XXX ...

Further  information:

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -9,24 +8,50 @@ RUN npm install
 COPY . .
 RUN npm run build
 
-FROM node:12.20.1-alpine3.9 as production
+FROM osgeo/gdal:alpine-normal-3.2.0 as production
+ENV CPL_VSIL_USE_TEMP_FILE_FOR_RANDOM_WRITE 'YES'
+RUN addgroup -g 1000 node \
+    && adduser -u 1000 -G node -s /bin/sh -D node
+#FROM node:14.18.2-alpine3.12 as production
```
Comment (asafMasa) [inline Dockerfile]: remove remarks

File: config/custom-environment-variables.json [hunk 2]
```diff
@@ -43,9 +43,23 @@
...
-    "format": "TILES_FORMAT"
+    "path": "TILES_PATH"
+  },
+  "vrt": {
+    "outputDirPath": "VRT_OUTPUT_DIR_PATH",
+    "sourcesListFilePath": "VRT_SOURCES_LIST_FILE_PATH",
+    "nodata": "VRT_NODATA",
+    "outputSRS": "VRT_OUTPUT_SRS",
+    "resampling": "VRT_RESAMPLING",
+    "addAlpha": "VRT_ADD_ALPHA"
```
Comment (asafMasa) [inline config/custom-environment-variables.json]: "__format": "boolean"

File: config/custom-environment-variables.json [hunk 3]
```diff
@@ -43,9 +43,23 @@
...
+  },
+  "generateTiles": {
+    "resampling": "GENERATE_TILES_RESMAPLING",
+    "profile": "GENERATE_TILES_PROFILE",
+    "srcnodata": "GENERATE_TILES_SRC_NODATA",
+    "verbose": "GENERATE_TILES_VERBOSE"
```
Comment (asafMasa) [inline config/custom-environment-variables.json]: "__format": "boolean"

File: requirements.txt [hunk 4]
```diff
@@ -0,0 +1,6 @@
+yarl
```
Comment (asafMasa) [inline requirements.txt]: why do we need this file?

File: src/tileSplitterManager.ts [hunk 5]
```diff
@@ -1,46 +1,58 @@
...
-      // const params = tilesTask.parameters as IParameters;
-      // const layerRelativePath = params.layerRelativePath;
+      const discreteId = (tilesTask.parameters as ITaskParams).discreteId;
+      const tilesDirectoryPath = this.tilesConfig.path;
+      const vrtPath = gdalUtilities.getVrtFilePath(discreteId);
+      const baseTilesPath = this.storageProvider === StorageProviderType.FS ? join(tilesDirectoryPath) : join(`/vsis3/${this.s3Config.bucket}`);
 
       if (attempts <= this.splitAttempts) {
         try {
           this.logger.info(`Running sync tiles task for taskId: ${tilesTask.id}, on jobId=${tilesTask.jobId}, attempt: ${attempts}`);
```
Comment (asafMasa) [inline src/tileSplitterManager.ts]: fix message - misspelling

File: src/tileSplitterManager.ts [hunk 6]
```diff
@@ -1,46 +1,58 @@
...
   public async handleSplitTilesTask(): Promise<boolean> {
     const tilesTask = await this.queueClient.queueHandlerForTileSplittingTasks.dequeue();
     if (tilesTask) {
+      const gdalUtilities = new GDALUtilities(this.logger, this.vrtConfig, this.generateTilesConfig, this.queueClient, tilesTask);
       const jobId = tilesTask.jobId;
       const taskId = tilesTask.id;
       const attempts = tilesTask.attempts;
-      // const params = tilesTask.parameters as IParameters;
-      // const layerRelativePath = params.layerRelativePath;
+      const discreteId = (tilesTask.parameters as ITaskParams).discreteId;
```
Comment (asafMasa) [inline src/tileSplitterManager.ts:35]: pass as parameter into GDALUtilities

File: src/gdalUtilities.ts [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+    private readonly generateTilesConfig: IGenerateTilesConfig,
+    private readonly queueClient: QueueClient,
+    private readonly task: ITaskResponse
+  ) {
+    this.jobId = this.task.jobId;
+    this.taskId = this.task.id;
+  }
+
+  public async buildVrt(task: ITaskResponse): Promise<void> {
+    const discreteId = (task.parameters as ITaskParams).discreteId;
```
Comment (asafMasa) [inline src/gdalUtilities.ts:30]: add buildvrt debug log

File: src/gdalUtilities.ts [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+      this.logger.error(`failed to create VRT file: ${error}`);
+      throw error;
+    } finally {
+      // removed the sources list file
+      await fsPromises.unlink(this.vrtConfig.sourcesListFilePath);
+    }
+  }
+
+  public async generateTiles(task: ITaskResponse, tilesBasePath: string): Promise<void> {
+    const discreteId = (task.parameters as ITaskParams).discreteId;
```
Comment (asafMasa) [inline src/gdalUtilities.ts:74]: add generateTiles debug log

**Review by asafMasa — CHANGES_REQUESTED** [review]
👍

File: Dockerfile (hunk 1 above)
Comment (CL-SHLOMIKONCHA) [reply Dockerfile]: fixed

File: config/custom-environment-variables.json (hunk 2 above)
Comment (CL-SHLOMIKONCHA) [reply config/custom-environment-variables.json]: fixed

File: requirements.txt (hunk 4 above)
Comment (CL-SHLOMIKONCHA) [reply requirements.txt]: unnecessary, removed

File: src/tileSplitterManager.ts (hunk 5 above)
Comment (CL-SHLOMIKONCHA) [reply src/tileSplitterManager.ts]: fixed

File: src/tileSplitterManager.ts (hunk 6 above)
Comment (CL-SHLOMIKONCHA) [reply src/tileSplitterManager.ts:35]: i've passed entire task to GDAL Utilities

File: src/gdalUtilities.ts (hunk 7 above)
Comment (CL-SHLOMIKONCHA) [reply src/gdalUtilities.ts:30]: fixed

File: src/gdalUtilities.ts (hunk 8 above)
Comment (CL-SHLOMIKONCHA) [reply src/gdalUtilities.ts:74]: fixed

File: src/tileSplitterManager.ts [hunk 9]
```diff
@@ -39,7 +39,7 @@ export class TileSplitterManager {
 
       if (attempts <= this.splitAttempts) {
         try {
-          this.logger.info(`Running sync tiles task for taskId: ${tilesTask.id}, on jobId=${tilesTask.jobId}, attempt: ${attempts}`);
+          this.logger.info(`Running split0 tiles task for taskId: ${tilesTask.id}, on jobId=${tilesTask.jobId}, attempt: ${attempts}`);
```
Comment (asafMasa) [inline src/tileSplitterManager.ts]: misspelling

File: config/custom-environment-variables.json (hunk 3 above)
Comment (CL-SHLOMIKONCHA) [reply config/custom-environment-variables.json]: fixed

**Review by asafMasa — APPROVED** [review]

**Review by Yossizz — APPROVED** [review]

**Pull request merged**